docs/specs/add-kubernetes-support.md

```diff
@@ -0,0 +1,247 @@
+single path. It will call the cluster object asking for an operation. For
+example, to create a new Pod:
+```diff
+- k8scli.CoreV1().Pods(ns).Create(pod)
++ Cluster.CreatePod(name, image, extraVars, pullPolicy, serviceAccountName, ns)
+```
+
+The Cluster object will recieve the request and call the correct API.
+```diff
++ func CreatePod(name string, image spec.Image, extraVars string,
++                pullPolicy string, serviceAccountName string, ns string) {
++	pod := &v1.Pod{
++		ObjectMeta: metav1.ObjectMeta{
++			Name: apbID,
++		},
++		Spec: v1.PodSpec{
++			Containers: []v1.Container{
++				{
++					Name:  "apb",
++					Image: spec.Image,
++					Args: []string{
++						action,
++						"--extra-vars",
++						extraVars,
++					},
++					ImagePullPolicy: pullPolicy,
++				},
++			},
++			RestartPolicy:      v1.RestartPolicyNever,
++			ServiceAccountName: serviceAccountName,
++		},
++	}
++
++       k8scli.CoreV1().Pods(ns).Create(pod)
++
++ }
+
+```
+
+If there are identical API objects between COEs, check which cluster is
+being used.
+```diff
++ func CreateClusterRoleBinding(roleBindingName string, svcAccountName string,
++                               ApbRole string, ns string) {
++	roleBindingM := map[string]interface{}{
++		"apiVersion": "v1",
++		"kind":       "RoleBinding",
++		"metadata": map[string]string{
++			"name":      roleBindingName,
++			"namespace": ns,
++		},
++		"subjects": []map[string]string{
++			map[string]string{
++				"kind":      "ServiceAccount",
++				"name":      svcAccountName,
++				"namespace": ns,
++			},
++		},
++		"roleRef": map[string]string{
++			"name": ApbRole,
++		},
++      }
++
++      if OpenShift {
++         OCcli.CoreV1().ClusterRoleBinding(ns).Create(roleBindingM)
++      } elif Kubernetes {
++         k8scli.CoreV1().ClusterRoleBinding(ns).Create(roleBindingM)
++      }
++ }
+
+```
+
+### Hierarchy Diagram
+The diagram represents the code split into two files and the API objects
+in each of them
+
+**pkg/cluster/kubernetes.go**
+
+| API Resource | Uses API | Uses CLI | Different between OpenShift and Kubernetes |
+|:---:|:---:|:---:|:---:|
+| Pods | X |  |  |
+| Exec |  | X |  |
+| Login |  | X |  |
+| ServiceAccount |  | X |  |
+| RBAC ClusterRoleBindings |  | X | X |
+| RBAC ClusterRoles |  | X | X |
+| Namespace |  | X | X |
+
+**pkg/cluster/openshift.go**
+
+| API Resource | Uses API | Uses CLI | Different between OpenShift and Kubernetes |
+|:---:|:---:|:---:|:---:|
+| ClusterRoleBindings |  | X | X |
+| ClusterRoles |  | X | X |
+| Project |  | X | X |
+
+## APBs
+The APBs only target OpenShift as a cluster. The APBs will differ in two ways:
+ - Packages
+ - oc-login.sh
+
+### Packages
+The main issue with packaging is that folks using Kubernetes may only want the
+Kubernetes client. The package ```origin-clients``` installs both the OpenShift
+and Kubernetes clients.  If this the container size isn't an issue, then there
+may not be a problem using ```origin-clients```.
+
```

Comment:
>But, that resource is mapped to the OpenShift security policy where as in Kubernetes it's Rbac. So I think it will work, but function different.

From my testing, RBAC entities are entirely ignored by OpenShift. It's very deceptive, because I was able to create RBAC *RoleBindings that appeared to have created absolutely fine (because they did), so my expectation was that they were, well...doing what role bindings do, but they were not from an openshift perspective. I am very hesitant to say this will work.

> Either way, I think we should discourage this behaviour and treat the playbooks and the clusters as completely independent entities.

+1. I think fundamentally we should encourage good design practice around this. Organize logic in such a way where shared logic can be passed around with things like roles, and encapsulate pieces that are specific to certain clusters.

WRT annotation, I would like to see something in the apb spec that tells the broker what COE a given APB supports. I think the broker should validate early, know what cluster it's *able* to deploy to and what APBs are *possible* to be deployed to that cluster. That way the broker can say `if !clusterSupports(apb) { handleUnsupported() } else { continue }`